### PR TITLE
Add option to show source references in justification tree

### PR DESCRIPTION
**Pull request by eshelyaron**
This PR propagates source references from the s(CASP) stack to the justification tree. Previously, source references were removed in `filter_tree/4`.
The `.pass` files with the expected justification trees for the test suite were updated accordingly.

Implications that appear in the human justification can now be annotated with references to the source definition of the relevant clause by passing the option `source(true)` to `human_justification_tree/2` or `html_justification_tree//2`, as in:
```
?- scasp(not flies(tweety), [tree(Tree)]), human_justification_tree(Tree, [source(true)]).
   there is no evidence that flies holds for tweety, because
      bird holds for tweety [per /checkouts/sCASP/test/all_programs/sasp/abdbirds.pl:10], and
      by abduction we conclude that ab holds for tweety [per /checkouts/sCASP/test/all_programs/sasp/abdbirds.pl:6]
   ∎
```

Source references are also added to the JSON output format, where tree nodes are annotated with `source_file` and `source_line` fields, like so:
```json
		  {
		    "children": [],
		    "node": {
		      "truth":true,
		      "value": {
			"type":"compound",
			"functor":"penguin",
			"args": [ {"type":"atom", "value":"tweety"} ]
		      },
		      "abduced":true,
		      "source_file":"/checkouts/sCASP/test/all_programs/sasp/abdbirds.pl",
		      "source_line":5
		    }
		  }
```

Source references are currently omitted from the "machine" output format, if needed I can those in a later PR.

### CODE REVIEW

File: prolog/scasp/dyncall.pl [hunk 1]
```diff
@@ -511,19 +535,21 @@
...
+abducible(Spec) :- abducible(Spec, false).
+
+abducible(M:(A,B), Pos) =>
+    abducible(M:A, Pos),
+    abducible(M:B, Pos).
+abducible(M:Head, Pos), callable(Head) =>
     abducible_rules(Head, Rules),
-    @(maplist(assertz, Rules), M).
+    maplist(scasp_assert_into(M, Pos), Rules),
+    @(assertz((:- discontiguous(('abducible$'/1, 'abducible$$'/1)))), M).
```
Comment (JanWielemaker) [inline prolog/scasp/dyncall.pl:539]: Why this change?   I've reverted as it does not work.  You cannot assert a directive; this will simply assert a rule for (:-)/1 (which should probably be forbidden).

File: prolog/scasp/html.pl [hunk 2]
```diff
@@ -8,8 +8,10 @@
           ]).
 :- use_module(common).
 :- use_module(output).
+:- use_module(dyncall).
```
Comment (JanWielemaker) [inline prolog/scasp/html.pl:11]: This one leads to a redefinition warning for not/1.    The reason isn't entirely clear to me.   We need to redefine not/1, but we can only do so if it is not yet imported into the current context.   It is redefined in embed.pl and then reexported to various modules.   This seems to go wrong on a cyclic import.   Anyway, as this library does not resolve anything for this module I removed the import.

I also added tests to `make check` to verify that the libraries and main project loads and passes basic checks.

File: prolog/scasp/dyncall.pl (hunk 1 above)
Comment (eshelyaron) [reply prolog/scasp/dyncall.pl:539]: >  You cannot assert a directive; this will simply assert a rule for (:-)/1 (which should probably be forbidden).

Thanks, that makes sense. Note that as far as I understand the issue you describe was in effect before this PR, we had:

```prolog
abducible(M:Head), callable(Head) =>
    abducible_rules(Head, Rules),
    @(maplist(assertz, Rules), M).

abducible_rules(HeadIn,
                [ (Head                 :- not AHead, 'abducible$'(Head)),
                  (AHead                :- not Head),
                  ('abducible$'(Head)   :- not 'abducible$$'(Head)),
                  ('abducible$$'(Head)  :- not 'abducible$'(Head)),
                  (:- discontiguous(('abducible$'/1, 'abducible$$'/1)))
                ]) :-
    ab_heads(HeadIn, Head, AHead).
```
So the term `:- discontiguous(('abducible$'/1, 'abducible$$'/1)` was passed as the argument of `assertz/1` it seems.

Because this PR added source positions for dynamic clauses, which should apply also to the clauses that make up an abducible predicate, I wanted to replace the plain `assertz/1`s with `scasp_assert/2` so `Pos` is asserted as the source position of the created clause. But `scasp_assert/2` (which `scasp_assert_into/3` delegates to) doesn't handle directives well, because it treats them as s(CASP) global constraints rather than Prolog directives, so I moved the `discontiguous` directive out of the `maplist` argument list in order to preserve the preexisting situation in which I figured it was passed to `assertz/1`.

Also note that current state with your latest fix may also not be correct, as because now we pass the `:- discontiguous/1` term to `scasp_assert_into/3` and as I mentioned above this results in the directive being treated as a global constraint.

Maybe we should just call `@(discontiguous(('abducible$'/1, 'abducible$$'/1)), M)` directly instead of asserting it? WDYT?

Comment (JanWielemaker) [reply prolog/scasp/dyncall.pl:539]: You are probably right.   The simple call is

    discontiguous(M:(...)).

No use for the complicated @/2 rules here.  Pushed 217a7185823dda891b08eac6b9c5ff316bc4b31c

**Comment by Gauntlet173** [comment]
FYI this PR seems to have caused a regression with regard to MQI

https://swi-prolog.discourse.group/t/change-in-scasp-breaks-mqi/5609

**Comment by eshelyaron** [comment]
Thanks @Gauntlet173, please see my reply on Discourse
